client/src/txnmgr.rs

```diff
@@ -0,0 +1,326 @@
+                    txid, height,
+                )
+            }
+        }
+    }
+}
+
+/// Transaction manager
+pub struct TxnManager<R: Reactor<Publisher>> {
+    handle: Handle<R>,
+}
+
+impl<R: Reactor<Publisher>> Clone for TxnManager<R>
+where
+    R::Waker: Sync,
+{
+    fn clone(&self) -> Self {
+        Self {
+            handle: self.handle.clone(),
+        }
+    }
+}
+
+impl<R: Reactor<Publisher>> TxnManager<R>
+where
+    R::Waker: Sync,
+{
+    /// Create a new TxnManager.
+    pub fn new(handle: Handle<R>) -> Self {
+        TxnManager { handle }
+    }
+
+    /// Submits a transaction.
+    ///
+    /// Transactions are broadcasted to as many connected peers as timeout duration
+    /// allows. To continue tracking transaction status after submission, do call
+    /// the wait method. This is useful if you want to track transactions till they
+    /// are accepted by peers.
+    /// ```ignore
+    /// let tx: Transaction;
+    /// let tx_id: tx.tx_id();
+    /// let txn_mgr = TxnManager::new(handle);
+    /// let status = txn_mgr.submit_transaction(tx).unwrap();
+    /// // Wait for transaction till accepted by a peer.
+    /// while let Ok(txnmgr::Event::Accepted {
+    ///     confirmations,
+    ///     txid,
+    /// }) = txn_mgr.wait(txn_id, std::time::Duration::from_secs(10))
+    /// {}
+    ///```
+    pub fn submit_transaction(
+        &mut self,
+        txn: Transaction,
+        timeout: time::Duration,
+    ) -> Result<Event, handle::Error> {
+        let tx_id = txn.txid();
+
+        let (transmitter, recvr) = chan::bounded(1);
+        self.handle
+            .command(Command::BroadcastTransactionInv(tx_id, transmitter))?;
+
+        let peers = recvr.recv()?;
+        if peers.is_empty() {
+            return Err(Error::RelayPeer.into());
+        }
+
+        let mut txn_map = match self.handle.transactions.write() {
+            Ok(e) => e,
+            Err(e) => return Err(e.into()),
+        };
+
+        let events = self.handle.events();
+        let mut peers_responded = HashSet::new();
+
+        // Keep on listening for peers that want our transaction till timeout.
+        let result = event::wait(
+            &events,
+            |e| match e {
+                event::Event::Received(e, NetworkMessage::GetData(inv)) => {
+                    if inv.contains(&Inventory::Transaction(tx_id)) {
+                        let result = self
+                            .handle
+                            .command(Command::SubmitTransaction(e, txn.clone()));
+                        peers_responded.insert(e);
+                        return Some(result);
+                    }
+
+                    None
+                }
+
+                event::Event::SyncManager(syncmgr::Event::BlockReceived(_, block, height)) => {
+                    self.on_blocks_received(tx_id, block, height, &mut *txn_map)
+                }
```

Comment:
It's better to rely on the existing high-level API for this, ie. `handle.blocks()`;

Reply:
Indeed, since they are filtered. I am working on a filtered event for transactions.

```rust
let (txn_filter_pub, txn_filter) = event::broadcast(|e| match e {
            Event::SyncManager(syncmgr::Event::BlockReceived(peerid, block, height)) => Some(
                Event::SyncManager(syncmgr::Event::BlockReceived(peerid, block, height)),
            ),

            Event::Received(e, NetworkMessage::GetData(inv)) => {
                Some(Event::Received(e, NetworkMessage::GetData(inv)))
            }

            Event::Received(e, NetworkMessage::Inv(inv)) => {
                Some(Event::Received(e, NetworkMessage::Inv(inv)))
            }

            _ => None,
        });
```